### PR TITLE
🐛 Story pages when removed makes navigation crash (WIP)

### LINKED ISSUE #29816: Removing a story page leads to console errors during development
To repro

- make a story with some pages
- load it in the browser
- remove the current page
- hit reload

Now all page loads of the URL start having an error in the console.

Before attempting to restore a page from history we should check whether that page still exists.

### PR DESCRIPTION
**Pull request by mszylkowski**
Fixes #29816

Steps to repro:
- Create a story and visualize it on a browser
- Remove the activePage from the story
- Refresh page

Page will not load since it cannot find the active page loaded from the browser history. There are two errors to fix: refreshing to a missing page, and advancing to a missing page.

Errors:
- `"Story refers to page "page-1", but no such page exists."` at `getPageDistanceMapHelper > getPageIndexById`
- `"Page at index  -1 exists, but is missing from the array."` at `getPageDistanceMapHelper > getPageById`

### CODE REVIEW

**Comment by mszylkowski** [comment]
Fixed the console errors associated with this bug, but the animations are still not working 100% **when you remove a page that is part of the navigation history**

removepage error

**Comment by gmajoulet** [comment]
We already have some code to work around that, we should build on top of it.
The issue is that in the `NAVIGATION_PATH` you have a page that no longer exists. Instead of doing a lot of DOM operations (getting the story, getting a page within that story) every time we navigate, let's just filter the `NAVIGATION_PATH` when we retrieve it from the browser history.

You can filter this array here and never have a wrong ID in the NAVIGATION_PATH:
https://github.com/ampproject/amphtml/blob/a7d7c97f12c491d0d886366b1f9e711e00850364/extensions/amp-story/1.0/amp-story.js#L2663-L2668

If you fix this TODO you can re-use this method and it won't even make any DOM queries:
https://github.com/ampproject/amphtml/blob/a7d7c97f12c491d0d886366b1f9e711e00850364/extensions/amp-story/1.0/amp-story.js#L1099-L1103

File: extensions/amp-story/1.0/amp-story.js [hunk 1]
```diff
@@ -2661,10 +2662,17 @@ export class AmpStory extends AMP.BaseElement {
...
-    this.storeService_.dispatch(
-      Action.SET_NAVIGATION_PATH,
-      getHistoryState(this.win, HistoryState.NAVIGATION_PATH) || []
+    let navigationPath = getHistoryState(
+      this.win,
+      HistoryState.NAVIGATION_PATH
     );
+    if (
+      !navigationPath ||
+      !navigationPath.every((pageId) => this.isActualPage_(pageId))
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js:2671]: Nit: you can exit the loop on the first non actual page

```suggestion
      navigationPath.some((pageId) => !this.isActualPage_(pageId))
```

**Review by gmajoulet — APPROVED** [review]

File: extensions/amp-story/1.0/amp-story.js (hunk 1 above)
Comment (mszylkowski) [reply extensions/amp-story/1.0/amp-story.js:2671]: https://jsfiddle.net/jqy431mh/1 `.every` also exits the loop on the first non-actual page

Comment (gmajoulet) [reply extensions/amp-story/1.0/amp-story.js:2671]: Makes sense 👍

**Pull request merged**